### PR TITLE
Creating new classes and associations

### PR DESCRIPTION
**Pull request by CarlosZ96**
Creating the Classroom, Book and Rental classes.
Creating the relationships between classes and the requested objects.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -66,3 +66,73 @@ def correct_name
...
+
+class Student
+  attr_accessor :name, :classroom
+  attr_reader :rentals
+
+  def initialize(name)
+    @name = name
+    @classroom = nil
+    @rentals = [] # Initialize the rentals array
+  end
```
Comment (rloterh) [inline main.rb]: - [ ] [OPTIONAL]
To adhere to best practices, consider organizing classes into separate files.
- [ ] [OPTIONAL]
To promote stronger adherence to object-oriented programming principles, let's establish inheritance from the Person class for the Student class.
- [ ] In the Student class, we want to establish the association that _a student belongs to a classroom_ . We can achieve that by doing something similar to:

```ruby
class Student < Person
  attr_reader :classroom
  .
  .

  def classroom(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
end
```

File: main.rb [hunk 2]
```diff
@@ -66,3 +66,73 @@ def correct_name
...
+end
+
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (rloterh) [inline main.rb]: - [ ] `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _write (set)_ it. Kindly ensure to improve upon this aspect.

File: main.rb [hunk 3]
```diff
@@ -66,3 +66,73 @@ def correct_name
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = [] # Initialize the rentals array
+  end
```
Comment (rloterh) [inline main.rb]: - [ ] As required, let's ensure that a method for adding rental is present in the Book class. The _book_ itself should be sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Book
  attr_accessor :title, :author, :rentals
  .
  .
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
end
```

File: main.rb [hunk 4]
```diff
@@ -66,3 +66,73 @@ def correct_name
...
+end
+
+class Rental
+  attr_accessor :date
+  attr_reader :book, :person
```
Comment (rloterh) [inline main.rb]: - [ ] Let's mark `book` and `person` with `attr_accessor`, as we intend to both _access (get)_ as well as _write (set)_ to it. Kindly ensure to improve upon this aspect.

File: main.rb [hunk 5]
```diff
@@ -66,3 +66,73 @@ def correct_name
...
+
+  def initialize(date, book, person)
+    @date = date
+    @book = book
+    @person = person
```
Comment (rloterh) [inline main.rb:118]: - [ ]  The Person class is missing. Kindly define a Person class which should be similar to:

```ruby
class Person
  attr_accessor :name, :age, :rentals
  .
  .

  def initialize(age, name = "Unknown", parent_permission = true)
    .
    .
    @rentals = []
  end

  def add_rental(book, date)
    Rental.new(date, book, self)
  end

  .
  .
end
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Carlos, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 
🏁 🚦 🏃 🌴 🏫 🌳 🏘️

## Highlights 👍  
- Nice job so far setting up associations!
- ReadMe looks professional!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

_Kindly check the inline comments marked **[OPTIONAL]** under the review._

**You can also consider:**
Let's make the ReadMe stand out even better by including an appropriate instruction under the _Getting Started_ section, on how to run and test the app, like `using rails server` command.

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: main.rb [hunk 6]
```diff
@@ -60,9 +59,73 @@ def correct_name
...
 puts capitalized_trimmed_person.correct_name
+
+class Student
+  attr_accessor :name, :classroom
+
+  def initialize(name)
+    @name = name
+    @classroom = nil
+  end
+end
```
Comment (Qoosim) [inline main.rb]: - [ ] This suggestion from the previous review has not been attended to :point_down: 

```
class Student < Person
  attr_reader :classroom
  .
  .

  def classroom(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
end
```
Kindly attend to it.

**Review by Qoosim — CHANGES_REQUESTED** [review]
#  *Status: Required Changes* :recycle: 

Hi @CarlosZ96, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


## Highlights
-  Implemented some required changes. :heavy_check_mark: 
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by OyePriscilla — APPROVED** [review]
Hi @CarlosZ96 👍 👍 👍, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights
Implemented method that serves as the 'belongs-to' side (a student belongs to a classroom) ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**